File: writeups/141_Linked_List_Cycle/ruby/main.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
  current_node = head
  checked_nodes = Set.new
  while current_node do
    if checked_nodes.include?(current_node)
      return true
    else
      checked_nodes.add(current_node)
      current_node = current_node.next
    end
  end
  
  false
end

File: writeups/141_Linked_List_Cycle/typescript/main.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
  let currentNode: ListNode = head;
  const checkedNodes: ListNode[] = [];
  while (currentNode) {
    if (checkedNodes.includes(currentNode)) {
      return true;
    } else {
      checkedNodes.push(currentNode);
      currentNode = currentNode.next;
    }
  }

  return false;
}
